refactor(transformer): pre-allocate more stack space

overlookmotel opened the pull request:
Pre-allocate 16 bytes for stacks in transforms that use them. Allocators usually support minimum of 16 bytes for allocations anyway. Now that we're using `SparseStack`, allocating decent capacity is now cheap.

overlookmotel commented:
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/6095?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#6095** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/6095?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈
* **#6094** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/6094?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#6093** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/6093?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#6092** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/6092?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#6091** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/6091?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#6090** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/6090?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#6089** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/6089?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @overlookmotel and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="11px" height="11px"/> <b>Graphite</b></a>
  <!-- Current dependencies on/for this PR: -->

overlookmotel commented:
This stack of PRs introducing `NonEmptyStack` and `Stack` has minimal perf impact (about +0.5% in total).

But we can also use these more efficient stack structures in many other places in our codebase, so the impact will be more significant in aggregate, especially for stacks which get pretty deep e.g. `TraverseAncestry`.

Dunqing commented on crates/oxc_transformer/src/es2015/arrow_functions.rs:
```diff
@@ -151,7 +151,7 @@ pub struct ArrowFunctions<'a> {
     pub fn new(options: ArrowFunctionsOptions) -> Self {
         // `SparseStack` is created with 1 empty entry, for `Program`
-        Self { _options: options, this_var_stack: SparseStack::new() }
+        Self { _options: options, this_var_stack: SparseStack::with_capacity(16, 0) }
```
Looks like we shouldn't do this in each plugin. Is it possible we use the `new` method and pre-allocate in it?

overlookmotel replied:
Yes, you're right. I've changed the default to 16 in `StackCapacity`, so it's applied everywhere. I don't understand why `std::vec::Vec` uses 8 as default.

Boshen reviewed (commented):
FYI previous PR crashed somewhere https://github.com/oxc-project/monitor-oxc/actions/runs/11065170732/job/30744171583

Running Transformer.
thread 'main' panicked at /home/runner/work/monitor-oxc/oxc/crates/oxc_transformer/src/helpers/stack/standard.rs:341:9:
assertion failed: self.cursor < self.end

overlookmotel committed:
refactor(transformer): pre-allocate more stack space (#6095)

Pre-allocate 16 bytes for stacks in transforms that use them. Allocators usually support minimum of 16 bytes for allocations anyway. Now that we're using `SparseStack`, allocating decent capacity is now cheap.
